dirdiff

```diff
@@ -0,0 +1,275 @@
+#!/usr/bin/env -S python
+# -*- coding: utf-8 -*-
+
+"""
+dirdiff [OPTIONS...] dir1 dir2
+modified files:
+1 column
+moved/renamed files:
+2 columns
+additional files:
+1 column
+"""
+
+from collections import defaultdict
+from contextlib import contextmanager
+from pathlib import Path
+from typing import Dict, List
+import argparse
+import dataclasses
+import hashlib
+import json
+import logging
+import os
+import stat
+
+LOG = logging.getLogger()
+LOG.setLevel("INFO")
+formatter = logging.Formatter("%(asctime)s - %(levelname)s - %(message)s")
+ch = logging.StreamHandler()
+ch.setFormatter(formatter)
+LOG.addHandler(ch)
+
+# types
+HexDigest = str
+
+
+@dataclasses.dataclass
+class DirDiffFile:
+    rootdir: Path
+    position: Path
+    hexdigest: str
+
+    def __init__(self, rootdir, path):
+        self.rootdir = rootdir
+        self.position = path
+        self.hexdigest = file_hexdigest(self.absolute)
+
+    @property
+    def absolute(self):
+        return self.rootdir / self.position
+
+
+class JSONEncoder(json.JSONEncoder):
+    def default(self, o):
+        if dataclasses.is_dataclass(o):
+            return dataclasses.asdict(o)
+        if isinstance(o, Path):
+            return str(o)
+        return super().default(o)
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+    )
+    parser.add_argument(
+        "-d",
+        "--debug",
+        action="store_true",
+        help="enable debug logging information",
+    )
+    parser.add_argument(
+        "--ignore-modified",
+        action="store_true",
+        help="don't show files that were modified",
+    )
+    parser.add_argument(
+        "--ignore-moved",
+        action="store_true",
+        help="don't show files that were moved around",
+    )
+    parser.add_argument(
+        "--ignore-additional",
+        action="store_true",
+        help="don't show files if they weren't either moved or modified",
+    )
+    parser.add_argument(
+        "--json",
+        action="store_true",
+        help="display output as JSON",
+    )
+    parser.add_argument(
+        "dirpath",
+        type=Path,
+        nargs="+",
+        help="dir path",
+    )
+    args = parser.parse_args()
+    if args.debug:
+        LOG.setLevel(logging.DEBUG)
+    return args
+
+
+@contextmanager
+def chdir(dirname):
+    oldpwd = Path().absolute()
+    try:
+        os.chdir(dirname)
+        yield
+    finally:
+        os.chdir(oldpwd)
+
+
+def apply_to_dir(dirpath, func):
+    return [
+        func(os.path.join(dirp, fname))
+        for dirp, _, filenames in os.walk(dirpath)
+        for fname in filenames
+    ]
+
+
+def file_hexdigest(filepath):
+    hasher = hashlib.sha1()
+    BUFSIZ = 65536
+    try:
+        with open(filepath, "rb") as f:
```

Comment:
I think this will not look at the file metadata (How do you want to proceed if you have the same file, with the same name but a different "modified at" etc...)

The problem is that this is OS specific and you may not want to deal with it

There's an example [here](http://timgolden.me.uk/python/win32_how_do_i/get-document-summary-info.html)

Reply:
Metadata can become messy real fast: do we consider identical files with changed modified_at (like with git) to be the same or different? Also, each filesystem will have different metadata so now the script may yield different results for the same directories setup on different filesystems.

For that reason, I decided to keep the script simple and straightforward in its role. If people want to compare metadata, they should find another way. I also suspect they'd want to check specific metadata values rather than everything